quill/ecs/src/component.rs

```diff
@@ -0,0 +1,67 @@
+use std::{alloc::Layout, any::TypeId, sync::Arc};
+
+use crate::space::MemorySpace;
+
+/// Type ID of a component.
+///
+/// Supports both Rust types and arbitrary
+/// "opaque" types identified by a `u64`.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct ComponentTypeId(ComponentTypeIdInner);
+
+impl ComponentTypeId {
+    pub fn of<T: 'static>() -> Self {
+        Self(ComponentTypeIdInner::Rust(TypeId::of::<T>()))
+    }
+
+    pub fn opaque(id: u64) -> Self {
```

Comment:
Maybe 32 bits would be enough?

Reply:
There's no harm in using a u64 so I disagree